### PR TITLE
[RLlib] NovelD exploration

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Sparse reward settings are very typical in RL, but need some more care in exploration. There are some algorithms in research that have made some progress on generating an intrinsic reward to help the agent to explore the environment faster. `Curiosity` is one amongst them and already implemented in `RLlib`.

In this PR I contribute `NovelD` (see [here](https://proceedings.neurips.cc/paper/2021/file/d428d070622e0f4363fceae11f4a3576-Paper.pdf) for the paper) another algorithm for fast and safe (more thoroughly) exploration based on a distillation network. 

In tests so far `NovelD` appears to perform similar to Curiosity, however in some settings it runs with better performance due to the fact that it curates only a single neural network (less than Curiosity) and reduces the intrinsic rewards drastically when the state has already been visited. 

Note, `NovelD` (as `Curiosity`) is not parallelized, yet. In a next step it is planned to produce an asynchronuous version of this algorithm to also make use of RLlib's scaling capabilities (@sven1977 Let's have a discussion on this). 

Note further that this algorithm (as also `Curiosity`) can only be used with `(Multi)Dicscrete` action spaces. Furthermore, if using continuous (or large discrete) state spaces in `NovelD`  the amount of memory grows as a list of hash codes for visited states is kept to calculate the intrinsic reward.  

This PR contains: 

1.  The NovelD algorithm.
2.  Some tests for the NovelD algorithm
3.  Metrics callbacks to keep track of exploration in TensorBoard and for the `tune.ExperimentAnalysis` 
4.  An example of how to use `NovelD` in `RLlib`.

@sven1977 @avnishn @gjoliver Let me know if we need some documentation somewhere

## Related issue number

None, it's a new feature. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
Hey @simonsays1980 ,
Looks great! I was going to ask how we evaluate that how well this implementation of NovelID works but I saw that you wrote a test that compares NovelID to curiosity.

File: rllib/examples/noveld_exploration.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+"""Example for using NovelD exploration for sparse reward settings.
+
+Note: NovelD works to this point only synchronously, i.e. the
```
Comment (ArturNiederfahrenhorst) [inline rllib/examples/noveld_exploration.py:3]: How can we change this? What do we need to make this work in distributed case?

Comment (simonsays1980) [reply rllib/examples/noveld_exploration.py:3]: @ArturNiederfahrenhorst I talked with @sven1977 about this and this should be the next step. In short, the updates of the NovelD network have to occur more often (namely in `postprocess_trajectory`. If now workers collect samples in parallel all these updates have to be synched. Theoretically, this will be something in in DDPPO. 
I am open for discussion (with a beer :))

**Comment by simonsays1980** [comment]
>  Yes, exactly. I know that Curiosity works on the examples and then comparing NovelD shows similar results at my side (hopefully also on yours :P)

**Comment by simonsays1980** [comment]
> Hey @simonsays1980 , Looks great! I was going to ask how we evaluate that how well this implementation of NovelID works but I saw that you wrote a test that compares NovelID to curiosity.

Yes, exactly. I know that Curiosity works on the examples and then comparing NovelD shows similar results at my side (hopefully also on yours :P)

File: rllib/examples/noveld_exploration.py (hunk 1 above)
Comment (sven1977) [reply rllib/examples/noveld_exploration.py:3]: Great question @ArturNiederfahrenhorst , and yeah, this is currently a somewhat unsolved problem: For curiosity + PPO, you need to more frequently update the ICM model (currently done inside the workers' postprocess trajectory after every(!) rollout) than the actual (e.g. PPO) policy model (only done when entire train batch is completely collected from all remote workers).

Once we have Ray Train running inside RLlib and therefore RLlib will have a simple "learner API" available in its `training_iteration` methods (and even on the rollout workers), then we can call these "sub-updates" of the ICM model easily from within the RolloutWorkers.

Sort of like this:
image

File: rllib/examples/noveld_exploration.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    env = gym.make(name)
+    env = gym_minigrid.wrappers.ImgObsWrapper(env)
+
+    return env
+
+
+register_env("mini-grid", env_creator)
+
+CONV_FILTERS = [[32, [5, 5], 1], [128, [3, 3], 2], [512, [4, 4], 1]]
+config = ppo.DEFAULT_CONFIG.copy()
```
Comment (sven1977) [inline rllib/examples/noveld_exploration.py:43]: Could we use the newest `PPOConfig()` class here?

```
from ray.rllib.agents.ppo import PPOConfig

config = PPOConfig().training(lr=..., model={...}).environment(env=...).framework(framework_str="torch").rollouts(num_envs_per_worker=4)

trainer = config.build()
```

File: rllib/utils/exploration/__init__.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 from ray.rllib.utils.exploration.exploration import Exploration
 from ray.rllib.utils.exploration.epsilon_greedy import EpsilonGreedy
 from ray.rllib.utils.exploration.gaussian_noise import GaussianNoise
+from ray.rllib.utils.exploration.noveld import NovelD
```
Comment (sven1977) [inline rllib/utils/exploration/__init__.py:5]: Awesome! :)

File: rllib/utils/exploration/noveld.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        random_timesteps: int = 10000,
+        sub_exploration: Optional[FromConfigSpec] = None,
+        **kwargs,
+    ):
+        """Initializes a NovelD exploration scheme.
+
+        Args:
+            action_space: The action space of the environment.
+            framework: The framework used to train the model.
+            model: The model used to train the model.
```
Comment (sven1977) [inline rllib/utils/exploration/noveld.py:76]: Could you complete this args-section? Missing the most useful stuff (how to configure the NovelD algo) :)

File: rllib/utils/exploration/noveld.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+
+        if self.policy_config["num_workers"] != 0:
+            raise ValueError(
+                "NovelD exploration currently does not support parallelism."
+                " `num_workers` must be 0!"
+            )
+
+        # Try to import xxhash.
+        try:
+            import xxhash
```
Comment (sven1977) [inline rllib/utils/exploration/noveld.py:93]: Nice!

File: rllib/utils/exploration/noveld.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+        self.intrinsic_reward_coeff = intrinsic_reward_coeff
+        self.normalize = normalize
+
+        self._state_counts = {}
+        self._state_counts_total = 0.0
+        self._state_counts_avg = 0.0
+        if self.normalize:
+            from ray.rllib.utils.exploration.random_encoder import MovingMeanStd
+
+            self._moving_mean_std = MovingMeanStd()
```
Comment (sven1977) [inline rllib/utils/exploration/noveld.py:128]: Can we set this to None, if self.normalize=False?

File: rllib/agents/callbacks.py [hunk 7]
```diff
@@ -568,3 +568,133 @@ def on_train_result(self, *, trainer, result: dict, **kwargs) -> None:
         # trainer._iteration as a parameter to on_learn_on_batch() call.
         RE3UpdateCallbacks._step = result["training_iteration"]
         super().on_train_result(trainer=trainer, result=result, **kwargs)
+
+
+class NovelDMetricsCallbacks(DefaultCallbacks):
```
Comment (sven1977) [inline rllib/agents/callbacks.py:573]: I think this should be part of the NovelD package. Could we move this class into utils/exploration/noveld.py?

File: rllib/examples/noveld_exploration.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        "type": "StochasticSampling",
+    },
+}
+# Use the callbacks for the NovelD metrics in TensorBoard.
+config["callbacks"] = NovelDMetricsCallbacks
+
+ray.init(local_mode=True, ignore_reinit_error=True)
+tune.run(
+    "PPO",
+    config=config,
```
Comment (sven1977) [inline rllib/examples/noveld_exploration.py:80]: if using the new config objects (tune does NOT support them yet) simply do `config=config.to_dict()` here.

**Review by sven1977 — COMMENTED** [review]
Looks really great @simonsays1980 , thanks for doing this!

Just a few comments and nits.

Also, in a follow-up PR, we should add this to the documentation (algorithm table) and write a few sentences about this algo, paper link, when to use it, is it better than curiosity, etc...

File: rllib/examples/noveld_exploration.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/examples/noveld_exploration.py:80]: Nice! Thanks for the tip!

File: rllib/utils/exploration/noveld.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/utils/exploration/noveld.py:128]: I have set the default to `self._moving_mean_std=None`

Comment (simonsays1980) [reply rllib/utils/exploration/noveld.py:128]: Done

File: rllib/agents/callbacks.py (hunk 7 above)
Comment (simonsays1980) [reply rllib/agents/callbacks.py:573]: > I think this should be part of the NovelD package. Could we move this class into utils/exploration/noveld.py?

I get some persistent circular imports here coming from the `DefaultCallbacks` import into `noveld.py` I guess

File: rllib/examples/noveld_exploration.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/examples/noveld_exploration.py]: Nice sketch @sven1977 - that makes it very clear. 

Waiting for `trainer_train` then ...

**Comment by simonsays1980** [comment]
@sven1977 Regarding documentation @christy gave me some hints. I will also ping @maxpumperla in case of questions and for review

File: rllib/utils/exploration/tests/test_noveld.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+                result = trainer.train()
+                print(result)
+                if result["episode_reward_max"] > 0.0:
+                    print("Reached goal after {} iters with NovelD!".format(i))
+                    learnt = True
+                    break
+            trainer.stop()
+            self.assertTrue(learnt)
+
+    def test_noveld_against_curiosity_on_partially_observable_domain(self):
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/utils/exploration/tests/test_noveld.py:224]: @sven1977 

- [ ] The first bullet point is easy to implement as far as I understand it and I can do it today or tomorrow. 
- [ ] The second bullet point I o not yet understand:
       - Do I take the actual tests and convert them to `tuned_examples`  with a `.yaml`?
       - Or do I take the new tests from the first bullet point and convert them to `tuned_examples`?

**Comment by ArturNiederfahrenhorst** [comment]
> > 
> 
> Thanks for the feedback @sven1977 - also helps to learn more about how you guys want to go with the code.
> 
> I would love to write some text about noveld. Who can guide me shortly how to change documentation?

If you want a general quick intro on how to contribute to docs, we can do a session. Ping if if that's it what you are looking for!

**Comment by simonsays1980** [comment]
> > > 
> > 
> > 
> > Thanks for the feedback @sven1977 - also helps to learn more about how you guys want to go with the code.
> > I would love to write some text about noveld. Who can guide me shortly how to change documentation?
> 
> If you just want a general quick intro on how to contribute to docs, we can do a session. So you can also ping me if if that is what you are looking for!

@ArturNiederfahrenhorst Yes would like to. Could also read the docs on how to do it, but I like to see best practices

**Comment by simonsays1980** [comment]
Due to #25292 the `test_noveld_metrics()` test will fail.

**Review by sven1977 — APPROVED** [review]
LGTM. Thanks for the PR @simonsays1980 , a huge value-add to RLlib!

**Comment by sven1977** [comment]
@simonsays1980 , let me quickly try to figure out the testing failures. If it's just circular dependencies and LINT stuff, should be easy to fix ...

**Comment by simonsays1980** [comment]
> 

The circular imports in the tests I have managed. The one in the `examples/noveld_exploration.py` is failing when executed. 

Also, let me know until which time you need the docs. Talked with @ArturNiederfahrenhorst for a meeting tomorrow to show me how to bring in the docs. Could also start a trial today.

File: rllib/utils/exploration/callbacks.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+from ray.rllib.agents.callbacks import DefaultCallbacks
+from ray.rllib.env.base_env import BaseEnv
+from ray.rllib.evaluation.episode import Episode
+
+if TYPE_CHECKING:
+    from ray.rllib.evaluation import RolloutWorker
+    from ray.rllib.policy import Policy
+
+
+class NovelDMetricsCallbacks(DefaultCallbacks):
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/exploration/callbacks.py:13]: NovelIDMetricsCallbacks?

**Comment by sven1977** [comment]
Hey @simonsays1980 , sorry for the delays. Let's merge this. Just checked the last non-passing test case, which was `test_nested_action_spaces`. It seems to pass fine on my laptop, but let's wait for everything to be green now, then merge today :)

File: doc/source/rllib/rllib-algorithms.rst [hunk 11]
```diff
@@ -62,6 +62,8 @@ Exploration-based plug-ins (can be combined with any algo)
 Algorithm                        Frameworks Discrete Actions        Continuous Actions Multi-Agent Model Support
 ================================ ========== ======================= ================== =========== =====================
 `Curiosity`_                     tf + torch **Yes** `+parametric`_  No                 **Yes**     `+RNN`_
+`NovelD`_                        tf + torch **Yes** `+parametric`_  Yes                **Yes**     `+RNN`_     
+`RE3`_                           tf         **Yes** `+parametric`_  Yes                **Yes**     `+RNN`_
```
Comment (sven1977) [inline doc/source/rllib/rllib-algorithms.rst:66]: Thanks for adding RE3 here as well! Very nice!

**Comment by sven1977** [comment]
Fixed LINTer (typo: "RLLib" -> "RLlib" :) )
Almost there ...
